api/pipeline_spec.proto

```diff
@@ -0,0 +1,420 @@
+  // Required field. The name of the pipeline.
+  // The name will be used to create or find pipeline context in MLMD.
+  string name = 1;
+}
+
+// The definition of a artifact type in MLMD.
+message ArtifactTypeSchema {
+  oneof kind {
+    // The name of the type. The format of the title must be:
+    // `<namespace>.<title>.<version>`.
+    // Examples:
+    //  - `aiplatform.Model.v1`
+    //  - `acme.CustomModel.v2`
+    // When this field is set, the type must be pre-registered in the MLMD
+    // store.
+    string schema_title = 1;
+
+    // Points to a YAML file stored on Google Cloud Storage describing the
+    // format.
+    string schema_uri = 2;
+
+    // Contains a raw YAML string, describing the format of
+    // the properties of the type.
+    string instance_schema = 3;
+  }
+}
+
+// The basic info of a task.
+message PipelineTaskInfo {
+  // The unique name of the task within the pipeline definition. This name
+  // will be used in downstream tasks to indicate task and data dependencies.
+  string name = 1;
+}
+
+// Definition for a value or reference to a runtime parameter. A
+// ValueOrRuntimeParameter instance can be either a field value that is
+// determined during compilation time, or a runtime parameter which will be
+// determined during runtime.
+message ValueOrRuntimeParameter {
+  oneof value {
+    // Constant value which is determined in compile time.
+    Value constant_value = 1;
+    // Name of the runtime parameter.
+    string runtime_parameter = 2;
+  }
+}
+
+// The definition of the deployment config of the pipeline. It contains the
+// the platform specific executor configs for KFP OSS.
+message PipelineDeploymentConfig {
+  // The specification on a container invocation.
+  // The string fields of the message support string based placeholder contract
+  // defined in [ExecutorInput](). The output of the container follows the
+  // contract of [ExecutorOutput]().
+  message PipelineContainerSpec {
+    // The image uri of the container.
+    string image = 1;
+    // The main entrypoint commands of the container to run. If not provided,
+    // fallback to use the entry point command defined in the container image.
+    repeated string command = 2;
+    // The arguments to pass into the main entrypoint of the container.
+    repeated string args = 3;
+
+    // The lifecycle hooks of the container.
+    // Each hook follows the same I/O contract as the main container entrypoint.
+    // See [ExecutorInput]() and [ExecutorOutput]() for details.
+    // (-- TODO(b/165323565): add more documentation on caching and lifecycle
+    // hooks. --)
+    message Lifecycle {
+      // The command and args to execute a program.
+      message Exec {
+        // The command of the exec program.
+        repeated string command = 2;
+        // The args of the exec program.
+        repeated string args = 3;
+      }
+      // This hook is invoked before caching check. It can change the properties

```

Comment:
How can this program change the properties of the execution and output artifacts?

Reply:
See comment on Lifecycle. We will use ExecutorOutput for output.